util/design/lib/OtpMemImg.py

```diff
@@ -0,0 +1,428 @@
+
+        # Key accounting
+        img_config_check = img_config.copy()
+        del img_config_check['seed']
+        del img_config_check['partitions']
+        _check_unused_keys(img_config_check, 'in image config')
+
+    def streamout_partition(self, part):
+        '''Scramble and stream out partition data as a list of bytes'''
+
+        part_name = part['name']
+        log.info('Streamout of partition {}'.format(part_name))
+
+        part_offset = part['offset']
+        part_size = part['size']
+        assert part_size % 8 == 0, 'Partition must be 64bit aligned'
+
+        # First chop up all items into individual bytes.
+        data_bytes = [0] * part_size
+        # Annotation is propagated into the hexfile as comments
+        annotation = ['unallocated'] * part_size
+        # Need to keep track of defined items for the scrambling.
+        # Undefined regions are left blank (0x0) in the memory.
+        defined = [False] * part_size
+        for item in part['items']:
+            for k in range(item['size']):
+                idx = item['offset'] - part_offset + k
+                annotation[idx] = part_name + ': ' + item['name']
+                if 'value' in item:
+                    data_bytes[idx] = ((item['value'] >> (8 * k)) & 0xFF)
+                    assert not defined[idx], "Unexpected item collision"
+                    defined[idx] = True
+
+        # Reshape this into 64bit blocks (this must be aligned at this point)
+        assert len(data_bytes) % 8 == 0, 'data_bytes must be 64bit aligned'
+
+        data_blocks = []
+        data_block_defined = []
+        for k, b in enumerate(data_bytes):
+            if (k % 8) == 0:
+                data_blocks.append(b)
+                data_block_defined.append(defined[k])
+            else:
+                data_blocks[k // 8] += (b << 8 * (k % 8))
+                # If any of the individual bytes are defined, the
+                # whole block is considered defined.
+                data_block_defined[k // 8] |= defined[k]
+
+        # Check if scrambling is needed
+        if part['secret']:
+            part_name = part['name']
+            key_sel = part['key_sel']
+            log.info('> Scramble partition with key "{}"'.format(key_sel))
+
+            for key in self.config['scrambling']['keys']:
+                if key['name'] == key_sel:
+                    break
+            else:
+                raise RuntimeError(
+                    'Scrambling key cannot be found {}'.format(key_sel))
+
+            for k in range(len(data_blocks)):
+                if data_block_defined[k]:
+                    data_blocks[k] = _present_64bit_encrypt(
+                        data_blocks[k], key['value'])
+
+        # Check if digest calculation is needed
+        if part['hw_digest']:
+            # Make sure that this HW-governed digest has not been
+            # overridden manually
+            if data_blocks[-1] != 0:
```

Comment:
Hey Michael, Sorry i do not understand why we need this logic here. I thought users can only input OTP content from the script, is there an entry that user can input digest directly from the hjson?
If so, why here we are only gating `hw_digests`?

Reply:
ah this is because we allow the user to override SW partition digests in the hjson.
the user can just add the correct item to then hjson file (e.g., `OWNER_SW_CFG_DIGEST`). 

but for partitions that have a HW governed digest, we disallow that, hence the check.

Reply:
Oh I see, thanks for explaining. Yeah it makes sense that script allow writing sw partitions digests!